sotopia/ui/fastapi_server.py

```diff
@@ -152,39 +205,58 @@ async def create_agent(agent: AgentProfileWrapper) -> str:
     return pk
 
 
-@app.post("/scenarios/", response_model=str)
-async def create_scenario(scenario: EnvironmentProfileWrapper) -> str:
-    scenario_profile = EnvironmentProfile(**scenario.model_dump())
-    scenario_profile.save()
-    pk = scenario_profile.pk
+@app.post("/relationship/", response_model=str)
+async def create_relationship(relationship: RelationshipWrapper) -> str:
+    relationship_profile = RelationshipProfile(**relationship.model_dump())
+    relationship_profile.save()
+    pk = relationship_profile.pk
     assert pk is not None
     return pk
 
 
-@app.put("/agents/{agent_id}", response_model=str)
-async def update_agent(agent_id: str, agent: AgentProfileWrapper) -> str:
-    try:
-        old_agent = AgentProfile.get(pk=agent_id)
-    except Exception:  # TODO Check the exception type
-        raise HTTPException(
-            status_code=404, detail=f"Agent with id={agent_id} not found"
-        )
-    old_agent.update(**agent.model_dump())  # type: ignore
-    assert old_agent.pk is not None
-    return old_agent.pk
-
-
-@app.put("/scenarios/{scenario_id}", response_model=str)
-async def update_scenario(scenario_id: str, scenario: EnvironmentProfileWrapper) -> str:
-    try:
-        old_scenario = EnvironmentProfile.get(pk=scenario_id)
-    except Exception:  # TODO Check the exception type
-        raise HTTPException(
-            status_code=404, detail=f"Scenario with id={scenario_id} not found"
-        )
-    old_scenario.update(**scenario.model_dump())  # type: ignore
-    assert old_scenario.pk is not None
-    return old_scenario.pk
+@app.post("/simulate/", response_model=str)
+async def simulate(simulate_request: SimulateRequest) -> str:
+    assert len(simulate_request.models) == 3, "There should be three models"
+    env_profile: EnvironmentProfile = EnvironmentProfile.get(pk=simulate_request.env_id)
+    env_params: dict[str, Any] = {
+        "model_name": simulate_request.models[0],
+        "action_order": "round-robin",
+        "evaluators": [
+            RuleBasedTerminatedEvaluator(
+                max_turn_number=simulate_request.max_turns, max_stale_turn=2
+            ),
+        ],
+        "terminal_evaluators": [
+            ReachGoalLLMEvaluator(
+                simulate_request.models[0],
+                EvaluationForTwoAgents[SotopiaDimensions],
+            ),
+        ],
+    }
+    env = ParallelSotopiaEnv(env_profile=env_profile, **env_params)
+    agents = Agents(
+        {
+            "agent1": LLMAgent(
+                "agent1",
+                model_name=simulate_request.models[1],
+                agent_profile=AgentProfile.get(pk=simulate_request.agent_ids[0]),
+            ),
+            "agent2": LLMAgent(
+                "agent2",
+                model_name=simulate_request.models[2],
+                agent_profile=AgentProfile.get(pk=simulate_request.agent_ids[1]),
+            ),
+        }
+    )
+    episode_pk = await arun_one_episode(
```

Comment:
It seems that this API would only return after the whole episode is run. A better design is to return an episode id immediately, and the user can check the episode's status via following calls.

Reply:
You add the `arun_one_episode` into background tasks.

Reply:
? Not quite sure if I follow here, are you referring to constantly saving episodes when each msg is produced?
> You add the arun_one_episode into background tasks.
what does this mean?

Reply:
So right now, the `episode_pk` is returned after the simulation is done. But this takes long to return to the client. So a better way is to return an `episode_pk` immediately, and run arun_one_episode in the background. Here is the pseudocode:

```python

episode_pk = EpisodeLog().pk
asyncio.create_task(arun_one_episode(..., episode_pk=episode_pk)) # update arun_one_episode to take an extra arg: episode_pk
return episode_pk

async def arun_one_episode(..., episode_pk: str):
    # run simulation
    ....
    # finish simulation
    episode_log.pk = episode_pk
    episode_log.save()

```

Reply:
A step after this is to add a simulation status:

```python
class SimulationStatus(JsonModel):
    episode_pk: str = Field(index)
    status: Literal["Started", "Error", "Completed"]

@app.get("/episodes/{episode_pk}")
class get_episode(episode_pk):
    status = SimulationStatus.find(SimulationStatus.episode_pk == episode_pk)
    if status.status in ["Started", "Error"]:
        return ....
    elif status.staus in ["Completed"]:
        return ...
```

Reply:
Updated the simulation status and made sure every simulation call would return an episode pk

I don't quite follow tho, about the `asyncio.create_task` one as users won't be able to check the episode until the task is done either way, and the current async implementation of the `async def simulate(simulate_request: SimulateRequest)` won't block any process?

Like people can still batch the `simulate` API call?

Reply:
Normally, when dealing with long-running requests, the standard way is to give client a fast response (202). Here are some references:
1. https://learn.microsoft.com/en-us/azure/architecture/patterns/async-request-reply
2. 202 (Accepted) as return code https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.3

Here is what Claude generates for me as an example workflow
```
POST /api/tasks
> 202 Accepted
{
    "task_id": "abc123",
    "status_url": "/api/tasks/abc123"
}

GET /api/tasks/abc123
> 200 OK
{
    "status": "processing",
    "progress": 45
}

GET /api/tasks/abc123
> 200 OK
{
    "status": "completed",
    "result": { ... }
}
```

Reply:
@ProKil I see,